src/Data/Functor/Foldable/TH.hs

```diff
@@ -70,12 +74,79 @@ import Paths_recursion_schemes (version)
 -- as we don't try to do better than
 -- <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#deriving-functor-instances GHC's DeriveFunctor>.
 --
-makeBaseFunctor :: Name -> DecsQ
-makeBaseFunctor = makeBaseFunctorWith baseRules
+-- For trickier data-types, like rose-tree (see also 'Cofree')
+--
+-- @
+-- data Rose f a = Rose a (f (Rose f a))
+-- @
+--
+-- we can invoke 'makeBaseFunctor' with instance declaration,
+-- to provide needed context for instances. (c.f. @StandaloneDeriving@)
+--
+-- @
+-- 'makeBaseFunctor' [d| instance Functor f => Recursive (Rose f a) |]
+-- @
+--
+-- Some doctests:
+--
+-- >>> data Expr a = Lit a | Add (Expr a) (Expr a) | Expr a :* [Expr a]
+-- >>> ; makeBaseFunctor ''Expr
+--
+-- >>> :t AddF
+-- AddF :: r -> r -> ExprF a r
+--
+-- >>> data Rose f a = Rose a (f (Rose f a))
+-- >>> ; makeBaseFunctor [d| instance Functor f => Recursive (Rose f a) |]
+--
+-- >>> :t RoseF
+-- RoseF :: a -> f r -> RoseF f a r
+--
+-- >>> let rose = Rose 1 (Just (Rose 2 (Just (Rose 3 Nothing))))
+-- >>> cata (\(RoseF x f) -> x + maybe 0 id f) rose
+-- 6
+--
+class MakeBaseFunctor a where
+    makeBaseFunctor :: a -> DecsQ
+
+instance MakeBaseFunctor a => MakeBaseFunctor [a] where
+    makeBaseFunctor a = fmap concat (T.mapM makeBaseFunctor a)
+
+instance MakeBaseFunctor a => MakeBaseFunctor (Q a) where
+    makeBaseFunctor a = makeBaseFunctor =<< a
+
+instance MakeBaseFunctor Name where
+    makeBaseFunctor = makeBaseFunctorWith baseRules
+
+-- | Expects declarations of 'Recursive' instances, e.g.
+--
+-- @
+-- makeBaseFunctor [d| instance Functor f => Recursive (Rose f a) |]
```

Comment:
Perhaps it is worth accepting either `Recursive` or `Corecursive` as the class name? After all, you're going to generate both.

Reply:
Yes. Done.

Reply:
Great! But the Haddocks still say "Expects declarations of `Recursive` instances". This should be broadened to include `Corecursive` as well.

Reply:
This still remains to be done, as far as I can tell.